RxSwift/Observables/Buffer.swift

```diff
@@ -136,3 +136,131 @@ final private class BufferTimeCountSink<Element, Observer: ObserverType>
+        return shared.buffer(trigger: shared.debounce(debounce, scheduler: scheduler))
+    }
+}
+
+final fileprivate class BufferTrigger<Element, TriggerElement> : Producer<[Element]> {
+    
+    fileprivate let _source: Observable<Element>
+    fileprivate let _trigger: Observable<TriggerElement>
+    
+    init(source: Observable<Element>, trigger: Observable<TriggerElement>) {
+        _source = source
+        _trigger = trigger
+    }
+    
+    override func run<O : ObserverType>(_ observer: O, cancel: Cancelable) -> (sink: Disposable, subscription: Disposable) where O.Element == [Element] {
+        let sink = BufferTriggerSink(parent: self, observer: observer, cancel: cancel)
+        let subscription = sink.run()
+        return (sink: sink, subscription: subscription)
+    }
+}
+
+final fileprivate class BufferTriggerSink<Element, TriggerElement, O: ObserverType>
+    : Sink<O>
+    , LockOwnerType
+    , ObserverType
+    , SynchronizedOnType where O.Element == [Element] {
+    typealias Parent = BufferTrigger<Element, TriggerElement>
+    typealias E = Element
+    
+    private let _parent: Parent
+    
+    let lock = RecursiveLock()
+    
+    // state
+    private let _serialDisposable = SerialDisposable()
+    private var _buffer = [E]()
+    
+    init(parent: Parent, observer: O, cancel: Cancelable) {
+        _parent = parent
+        super.init(observer: observer, cancel: cancel)
+    }
+    
+    func run() -> Disposable {
+        let disposable = SingleAssignmentDisposable()
+        _serialDisposable.disposable = disposable
+        disposable.setDisposable(_parent._trigger.subscribe { (event: Event<TriggerElement>) in
+            switch event {
+            case .next(_):
+                self.startNewWindowAndSendCurrentOne()
+                break
+            case .error(let error):
+                self._buffer = []
+                self.forwardOn(.error(error))
+                self.dispose()
+                break
+            case .completed:
+                if self._buffer.count > 0 {
+                    self.forwardOn(.next(self._buffer))
+                }
+                self.forwardOn(.completed)
+                self.dispose()
+                break
+            }
+        })
+        
+        return Disposables.create(_serialDisposable, _parent._source.subscribe(self))
+    }
+    
+    func startNewWindowAndSendCurrentOne() {
+        let buffer = _buffer
+        _buffer = []
+        if buffer.count > 0 {
+            forwardOn(.next(buffer))
+        }
+    }
+    
+    func on(_ event: Event<E>) {
+        synchronizedOn(event)
+    }
+    
+    func synchronized_on(_ event: Event<Element>) {
+        switch event {
+        case .next(let element):
+            _buffer.append(element)
+        case .error(let error):
+            _buffer = []
+            forwardOn(.error(error))
+            dispose()
+        case .completed:
+            if _buffer.count > 0 {
```

Comment:
```suggestion
            if !buffer.isEmpty {
```

Preferable over explicit count check